test/h5test.h

```diff
@@ -101,79 +143,101 @@ H5TEST_DLLVAR MPI_Info h5_io_info_g; /* MPI INFO object for IO */
  */
 #define TESTING(WHAT)                                                                                        \
     do {                                                                                                     \
-        printf("Testing %-62s", WHAT);                                                                       \
-        fflush(stdout);                                                                                      \
-        n_tests_run_g++;                                                                                     \
+        if (IS_MAIN_TEST_THREAD) {                                                                           \
+            printf("Testing %-62s", WHAT);                                                                       \
+            fflush(stdout);                                                                                      \
+        }                                                                                                   \
+        INCR_TEST_STAT(n_tests_run_g); \
```

Comment:
Since threads will be sharing the same variables here, any increments to these globals should only be done from a single thread and not from all of them.

Reply:
Yes, the same as for the parallel tests, I don't think it makes sense for the number of tests to be different just because multiple threads are running the same test. Test failures in non-main threads should be reported back to the main thread at some point with some kind of reduction operation, at which point the global `num_errs_g` can be set appropriately (though we should consider how a single, global error count fits into multi-threaded tests). I don't think we necessarily need a wrapper around the tests either. It seems like the return value of `PerformThreadTest` can simply be set to the return value of the test. Then it can be captured from each thread during `pthread_join` and coalesced. There might be an even better design, but we're starting to get to the point of challenges with shoehorning multi-thread functionality into originally non-multi-threaded tests.

Reply:
The individual tests don't have return values since their integration with testframe. I'll try an initial implementation using results stored on the threadlocal info visible to both the API tests and testframe.

Reply:
I've got an implementation that aggregates results across runs of the same test in different threads, but sub-tests present an issue. Testframe isn't aware of sub-tests and needs some kind of additional information in order to correctly report a situation where e.g. a single sub-test fails in a single thread, but the rest all pass. 

I think that a threadlocal array of sub-test results with subtests identified by index would be enough for testframe for aggregate results in the way we'd like. 

Eventually maybe sub-tests could be reworked into full tests, but that would be an absolutely massive change.

Reply:
Let's go ahead and have that separately since getting this PR in is somewhat blocking both of us